cura/NetworkClient.py

```diff
@@ -0,0 +1,220 @@
+
+        # QHttpMultiPart objects need to be kept alive and not garbage collected during the
+        # HTTP which uses them. We hold references to these QHttpMultiPart objects here.
+        self._kept_alive_multiparts = {}  # type: Dict[QNetworkReply, QHttpMultiPart]
+
+    ##  Creates a network manager with all the required properties and event bindings.
+    def _createNetworkManager(self) -> None:
+        if self._manager:
+            self._manager.finished.disconnect(self.__handleOnFinished)
+            self._manager.authenticationRequired.disconnect(self._onAuthenticationRequired)
+        self._manager = QNetworkAccessManager()
+        self._manager.finished.connect(self.__handleOnFinished)
+        self._last_manager_create_time = time()
+        self._manager.authenticationRequired.connect(self._onAuthenticationRequired)
+
+    ##  Create a new empty network request.
+    #   Automatically adds the required HTTP headers.
+    def _createEmptyRequest(self, url: str, content_type: Optional[str] = "application/json") -> QNetworkRequest:
+        request = QNetworkRequest(QUrl(url))
+        if content_type:
+            request.setHeader(QNetworkRequest.ContentTypeHeader, content_type)
+        request.setHeader(QNetworkRequest.UserAgentHeader, self._user_agent)
+        return request
+
+    ##  Executes the correct callback method when a network request finishes.
+    def __handleOnFinished(self, reply: QNetworkReply) -> None:
+        
+        # Due to garbage collection, we need to cache certain bits of post operations.
+        # As we don't want to keep them around forever, delete them if we get a reply.
+        if reply.operation() == QNetworkAccessManager.PostOperation:
+            self._clearCachedMultiPart(reply)
+
+        # No status code means it never even reached remote.
+        if reply.attribute(QNetworkRequest.HttpStatusCodeAttribute) is None:
+            return
+
+        # Not used by this class itself, but children might need it for better network handling.
+        # An example of this is the _update method in the NetworkedPrinterOutputDevice.
+        self._last_response_time = time()
+        
+        # Find the right callback and execute it.
+        # It always takes the full reply as single parameter.
+        callback_key = reply.url().toString() + str(reply.operation())
+        if callback_key in self._on_finished_callbacks:
+            self._on_finished_callbacks[callback_key](reply)
+
+    ##  Removes all cached Multi-Part items.
+    def _clearCachedMultiPart(self, reply: QNetworkReply) -> None:
+        if reply in self._kept_alive_multiparts:
+            del self._kept_alive_multiparts[reply]
+
+    ##  Makes sure the network manager is created.
+    def _validateManager(self) -> None:
+        if self._manager is None:
+            self._createNetworkManager()
+        assert (self._manager is not None)
```

Comment:
Superfluous ()

Reply:
on older versions of Osx (not sure if it's still the case), you need to re-create the networkManager if the computer went into sleep and got out of it again (otherwise it wouldn't be able to send / receive messages at all)

Reply:
Ah okay. Will keep it this way then.